### PR TITLE
Issue #185: ensure uniqueness of request ids to prevent incorrect response to handler assignment

### LINKED ISSUE #185: I get ua.BrowseResponse on ua.PublishRequest
Goroutine 1: Sending PublishRequest and waiting for response
Goroutine 2: Sending BrowseRequest and waiting for response
Goroutine 1: Received ua.BrowseResponse
Goroutine 2: waiting

Seems like related to SequenceNumber in #184

### PR DESCRIPTION
**Pull request by alexbrdn**
- increase `SecureChannel.cfg.RequestId` on each request, otherwise all requests have the same request id and the responses to parallel requests may get assigned incorrectly
- drop decrementing the counters:
  * multiple parallel requests may get the same ids
  * SequenceId is supposed to be monotonically rising anyways
- split and reduce the block protected by mutex (see #186). There's no need to encode the message and write to secure channel in protected block
- add a fail-safe check to prevent overriding response handler if there is already a handler waiting for response with the same id

closes gopcua/opcua#185

### CODE REVIEW

**Comment by alexbrdn** [comment]
@magiconair, as I am very new to the matter, I appreciate your feedback. My understanding of the existing code and the problem at hand might be way off, so don't hesitate to tear it apart.

File: uasc/secure_channel.go [hunk 1]
```diff
@@ -267,9 +261,12 @@ func (s *SecureChannel) SendAsync(svc interface{}, authToken *ua.NodeID) (resp c
 
 	// register the handler
 	resp = make(chan Response)
-	// s.mu.Lock()
+	s.mu.Lock()
+	if _, ok := s.handler[reqid]; ok {
```
Comment (magiconair) [inline uasc/secure_channel.go]: you can simplify this to 

```go
if s.handler[reqid] != nil {
    return nil, fmt.Errorf("error: duplicate handler registration for request id %d", reqid)
}

**Review by magiconair — COMMENTED** [review]
Small improvement. Otherwise, LGTM

**Comment by magiconair** [comment]
Also, that approach makes sense. Thank you!

**Comment by magiconair** [comment]
But that would mean that sending some of your requests had to generate an error.

File: uasc/secure_channel.go (hunk 1 above)
Comment (alexbrdn) [reply uasc/secure_channel.go]: good point. changed it

**Comment by alexbrdn** [comment]
>But that would mean that sending some of your requests had to generate an error.

No, none generated error. It's just that there were two requests sent in parallel and the handler got overwritten. So there were two goroutines waiting for response and only one handler. If the incorrect response arrived first, it produced an error. If the correct arrived first, then the second goroutine was left waiting forever for a response.

**Comment by magiconair** [comment]
I don't see how that would be possible since `SendAsync` holds the lock for the duration of the call including updating the counters and the map. How can two go routines get the same request id unless the first one fails and resets the counter. 

I've also just noticed that we're not handling the the error for 

```go
b, err = s.signAndEncrypt(m, b)
```

Could you please add that as well?

**Comment by magiconair** [comment]
Without the lock from #186 this was possible but with the lock ...

**Comment by magiconair** [comment]
I assume that with your fix you don't see the error anymore.

**Comment by alexbrdn** [comment]
the `RequestId` is assigned from Config and it was never updated before this change. So _all_ requests had the same id. Only `SequenceNumber` and `RequestHandle` were incremented/decremented.

I'll add an error check.

**Comment by alexbrdn** [comment]
That's true, I don't get the error anymore. I think I'll be able to release the code that produced the error as part of `examples/subscribe/subscribe.go` next week. I'm working on subscription functionality and this was a stumbling block.

**Comment by magiconair** [comment]
Subscription would be awesome! Thanks for digging into this.

**Comment by magiconair** [comment]
Aaaah, now I see it. duh. So the examples were too simple. :)

**Pull request merged**